### PR TITLE
[WIP] Add `scatter_to_table`

### PR DESCRIPTION
**Pull request by shwina**
Adds a `scatter_to_table` API (unrelated to the previous, and now dropped, libcudf function `scatter_to_tables`) for scattering values from a column into a table, based on the provided row and column indices.

Together with #4574, this will help implement #1821.

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

**Comment by shwina** [comment]
@jrhemstad pointed out that this can be achieved by doing multiple scatters:

1. Sort the columns `input, row_indices, column_indices` by `column_indices`. 
2. Split the result from step 1 at the appropriate locations so as to get the scatter map for each output column.
3. Perform `N` scatters from the input column to each of the `N` columns of the result table based on the scatter maps generated in step 2.

Some potential issues with this:

* It's not immediately clear how to determine the splits in step 2
* When `N` is large, it might be much more efficient to do this using a single kernel

**Comment by jrhemstad** [comment]
After discussing with @shwina, we realized that this is basically converting a COO coordinate list to a `cudf::table`.

**Comment by shwina** [comment]
So I realized today that `scatter_to_table` can be implemented essentially with a single scatter.


row | column | value
----|---------|------
1 | 3 | 9
2 | 2 | 8
0 | 1 | 2
0 | 0 | 1
1 | 2 | 4
2 | 1 | 6

Expect:

col0 | col1 | col2 | col3
-----|-----|------|----
1 | 2 | null | null
null | null | 4 | 9
null | 6 | 8 | null

```
num_output_rows = max(row) + 1 = 3
num_output_columns = max(column) + 1 = 4
```

Convert the `(row, value)` pairs to a scatter map:

```
scatter_map[i] = column[i] * num_rows + row[i]
scatter_map == [10, 8, 3, 0, 7, 5]
```

Conver the scatter map to a gather map of size (num_output_rows + num_output_columns):

```
gather_map == [3, -1, -1, 2, -1, 5, -1, 4, 1, -1, 0, -1, -1]
value == [9, 8, 2, 1, 4, 6]
```

Gather `value` using `gather_map`:

```
result == [1, null, null, 2, null, 6, null, 4, 8, null, 9, null]
```

Reshape result into a table of `num_output_rows` and `num_output_columns`:

```
Got:

col0 | col1 | col2 | col3
-----|-----|------|----
1 | 2 | null | null
null | null | 4 | 9
null | 6 | 8 | null
```

**Comment by shwina** [comment]
@jrhemstad At this point, I'm wondering if it's worth having a libcudf API for this.

File: cpp/src/copying/scatter_to_table.cu [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+                                                                MapIterator row_labels_end,
+                                                                MapIterator column_labels_begin,
+                                                                MapIterator column_labels_end,
+                                                                size_type num_output_rows,
+                                                                size_type num_output_columns,
+                                                                rmm::mr::device_memory_resource* mr,
+                                                                cudaStream_t stream)
+{
+  auto scatter_to_table_functor = column_to_table_scatterer<decltype(row_labels_begin)>{};
+  return type_dispatcher(input.type(),
```
Comment (jrhemstad) [inline cpp/src/copying/scatter_to_table.cu:100]: Why do you need to dispatch on the `input` type here? When you later call `cudf::scatter` it's just going to have to do the same dispatch again.

File: cpp/src/copying/scatter_to_table.cu [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+    return column_labels_begin[i] * num_output_rows + row_labels_begin[i];
+  }
+
+  MapIterator row_labels_begin, column_labels_begin;
+  size_type num_output_rows;
+};
+
+template <typename MapIterator>
+struct column_to_table_scatterer {
+  template <typename Element>
```
Comment (jrhemstad) [inline cpp/src/copying/scatter_to_table.cu:40]: As I suspected, `Element` isn't actually used, so this dispatch is unnecessary.

**Comment by shwina** [comment]
Results of benchmarking `pivot` based on libcudf `scatter_to_tables` , v/s a cudf-only `scatter_to_tables` based on [this algorithm](https://github.com/rapidsai/cudf/pull/5448#issuecomment-647824024).

libcudf:

```
size: 100 :: time: 0.04611291101900861
size: 200 :: time: 0.057594457000959665
size: 500 :: time: 0.1314852550276555
size: 1000 :: time: 0.2992417369969189
size: 5000 :: time: 2.875286730006337
size: 10000 :: time: 10.198028729006182
```

cudf:

```
size: 100 :: time: 0.05878497299272567
size: 200 :: time: 0.0836187910172157
size: 500 :: time: 0.22112199204275385
size: 1000 :: time: 0.41312640998512506
size: 5000 :: time: 3.829778195999097
size: 10000 :: time: 12.141353156999685
```

The benchmark is essentially constructing a square matrix from COO format co-ordinates:

```python
import cupy
import cudf
import timeit

for size in [100, 200, 500, 1_000, 5_000, 10_000]:
    columns, index = cupy.meshgrid(cupy.arange(size), cupy.arange(size))
    df = cudf.DataFrame({
        'index': index.ravel(),
        'columns': columns.ravel(),
        'values': cupy.arange(size**2)
    })
    start = timeit.default_timer()
    for i in range(10):
        result = df.pivot('index', 'columns')
    end = timeit.default_timer()
    print(f"size: {size} :: time: {end-start}")
```

**Comment by shwina** [comment]
Based on these results, it's probably not worth adding a new libcudf API for this. Closing.

**Pull request closed without merging**